src/PSR7/Validators/BodyValidator/MultipartValidation.php

```diff
@@ -0,0 +1,187 @@
+
+use cebe\openapi\spec\Encoding;
+use cebe\openapi\spec\Header;
+use cebe\openapi\spec\MediaType;
+use cebe\openapi\spec\Schema;
+use cebe\openapi\spec\Type as CebeType;
+use OpenAPIValidation\PSR7\Exception\Validation\InvalidBody;
+use OpenAPIValidation\PSR7\Exception\Validation\InvalidHeaders;
+use OpenAPIValidation\PSR7\OperationAddress;
+use OpenAPIValidation\Schema\Exception\SchemaMismatch;
+use OpenAPIValidation\Schema\Exception\TypeMismatch;
+use OpenAPIValidation\Schema\SchemaValidator;
+use Psr\Http\Message\MessageInterface;
+use Riverline\MultiPartParser\Converters\PSR7;
+use Riverline\MultiPartParser\StreamedPart;
+use RuntimeException;
+use const JSON_ERROR_NONE;
+use function in_array;
+use function json_decode;
+use function json_last_error;
+use function json_last_error_msg;
+use function preg_match;
+use function sprintf;
+use function str_replace;
+use function strpos;
+
+/**
+ * Should validate multipart/* body types
+ */
+trait MultipartValidation
+{
+    /**
+     * @see https://swagger.io/docs/specification/describing-request-body/multipart-requests/
+     * @see https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#requestBodyObject
+     *
+     * @param MediaType[] $mediaTypeSpecs
+     */
+    private function validateMultipart(OperationAddress $addr, MessageInterface $message, array $mediaTypeSpecs, string $contentType) : void
+    {
+        /** @var Schema $schema */
+        $schema = $mediaTypeSpecs[$contentType]->schema;
+
+        // 0. Multipart body message MUST be described with a set of object properties
+        if ($schema->type !== CebeType::OBJECT) {
+            throw TypeMismatch::becauseTypeDoesNotMatch('object', $schema->type);
+        }
+
+        // 1. Parse message body
+        $document = PSR7::convert($message);
+
+        // 2. Validate bodies of each part
+        $body = $this->parseMultipartData($addr, $document);
+
+        $validator = new SchemaValidator($this->detectValidationStrategy($message));
+        try {
+            $validator->validate($body, $schema);
+        } catch (SchemaMismatch $e) {
+            throw InvalidBody::becauseBodyDoesNotMatchSchema($contentType, $addr, $e);
+        }
+
+        // 3. Validate specified part encodings and headers
+        // @see https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#encoding-object
+        // The encoding object SHALL only apply to requestBody objects when the media type is multipart or application/x-www-form-urlencoded.
+        // An encoding attribute is introduced to give you control over the serialization of parts of multipart request bodies.
+        // This attribute is only applicable to "multipart" and "application/x-www-form-urlencoded" request bodies.
+        $encodings = $mediaTypeSpecs[$contentType]->encoding;
+
+        foreach ($encodings as $partName => $encoding) {
+            $parts = $document->getPartsByName($partName); // multiple parts share a name?
```

Comment:
anwering comment question, I think this could happen when dealing of data arrays. than name could be identical and\or like multiple `name[]`